[yadobler] iP

yadobler opened the pull request:
# YappingBot
> _"Software is like s-x: it's better when it's free."_ - Linus Torvalds ([Source](https://x.com/Linus__Torvalds/status/296333253571387392), yes he actually said that.)

YappingBot **FREES** your mind of having to remember things you need to do.

- Text-based
- Questionable learning curve
- _SUPER_ FAST to ~~crash~~ use (I'm kidding, it's Java. You only get that* in C)
  (_* - in both speed and stability aspects_)

All you need to do is:
1. Download it from [here](https://github.com/yadobler/ip/releases) (DID YOU CHECK THE SITE LINK?).
2. Double click it (honestly doesn't work for me. You're better off running `java -jar yappingbot.java`).
3. ???
4. Add your tasks!

Supported flags include:
```
--savefile <FILEPATH> define an alternate savefile path
        -s <FILEPATH> same as --savefile
        -c            Run CLI mode (fallback)
```

Did I mention it's **FREE**? (Free of Charge, Free of Dom)

# Features implemented:
- [x] View all added tasks
- [x] Manages _to-do_ tasks
- [x] Manages _deadline_ tasks
- [x] Manages _event_ tasks
- [x] Task list persistence (automatic save to and retrieve from disk)
- [ ] Does your taxes
- [x] Search and filter tasks (iteratively too!)
- [x] GUI
- [x] CLI mode fallback (pass `-c`)
- [x] Change savefile path (pass `-s <filepath>`)
- [X] Add custom aliases
- [ ] Add CSS to make things pretty
- [ ] Add more personality
- [ ] Persistant settings (ie for aliases)
- [ ] Autosave
- [ ] Flexible date formatting
- [ ] More coming soon...
---
If you're bored and want to stress yourself, look no further!

Clone this, and feel free to run it like this:

**CLI Mode:**
```java
public static void main(String[] args) {
    // Feel free to change this!
    // The original main method checks if an argument is passed and uses it instead.
    String savefile = "./savefile"; 

    YappingBot yp = new YappingBot(savefile);
    yp.start();
}
```

**GUI Mode:** (requires javafx)
```java
public static void main(String[] args) {
    // MainGuiApplication currently looks in 
    // Launcher to retrieve the savefilePath
    //
    // To be updated in further commits
    Launcher.savefilePath = "./savefile";
    MainGuiApplication.launch(MainGuiApplication.class, args);
}
```

yadobler committed:
Add deserialiaze function for tasks

yadobler committed:
Modified deserialiaze to accept string array instead

yadobler committed:
Added constructor with no arguments for tasks

yadobler committed:
Add error message for no save file

yadobler committed:
Implemented save feature

yadobler committed:
Typo: Implement LOAD feature, save feature not implemented

yadobler committed:
Implement save feature; update .gitignore to exclude savefile

yadobler committed:
Fix bugs with saving and loading and serialization

yadobler committed:
Fix typo in function name

yadobler committed:
Merge branch 'branch-Level-7'

yadobler committed:
Fix datetime

yadobler committed:
Change LocalDateTime to LocalDate

yadobler committed:
Merge branch 'branch-Level-8'

yadobler committed:
Add Ui class

yadobler committed:
Rename Commands.java to CommandTypes.java;Moved to commands package

yadobler committed:
Ongoing OOP

yadobler committed:
Fixed ui

yadobler committed:
Update runtest.sh to test loading and saving

yadobler committed:
Update runtest.sh to test error in savefile

yadobler committed:
Add savefile_original

yadobler committed:
Add Storage class

yadobler committed:
Add argument option for savefile path

yadobler committed:
Fix stuff

yadobler committed:
Move parsing command into Parser class

yadobler committed:
Add TaskList class; Move intParser from commands to Parser; Move tasklist-related commands to Parser

yadobler committed:
Implement Iterable for Tasklist class

yadobler committed:
Merge branch 'branch-A-MoreOOP'

yadobler committed:
Refactor

yadobler committed:
Merge branch 'add-gradle-support'

yadobler committed:
Add root package

yadobler committed:
ditto

yadobler committed:
gradle works

yadobler committed:
Fix packaging issues

yadobler committed:
Generate jUnit test files for ui package

yadobler committed:
Generate jUnit test files for tasks package

yadobler committed:
Generate jUnit test files for commands package

yadobler committed:
Generate jUnit test files and Resource (savefile) for storage

yadobler committed:
Undo incorrect jUnit test generated files

yadobler committed:
Generate test file for storage class

yadobler committed:
Add test for Storage.loadListFromFile

yadobler committed:
Add test for Storage.saveListFromFile

yadobler committed:
Fix typo 'testSaveListFromFile' -> 'testSaveListToFile'

yadobler committed:
Add extra ui test for Storage.loadListFromFile

yadobler committed:
Add savefile_test output to .gitignore

yadobler committed:
Create tasks.TaskParser; Move taskParser from Storage to TaskParser

yadobler committed:
Move methods around to encapsulate task parsing to only tasks package

yadobler committed:
Fix runtest.sh search path

yadobler committed:
Fix jUnit tests to isolate class

yadobler committed:
Add jUnit test for Ui.quoteSingleLine

yadobler committed:
Add jUnit test for ui.Ui

yadobler committed:
Remove unused imports in UiTest.java

yadobler committed:
Merge branch 'branch-A-JUnit'

yadobler committed:
Add checkstyle plugin to gradle

yadobler committed:
Fix typo in expected output

yadobler committed:
Fix wrong line order in testSaveListToFile

yadobler committed:
Remove unused throws and imports

yadobler committed:
Remove checkstyle for now

yadobler committed:
Fix gradle build for jar

yadobler committed:
Merge branch 'branch-A-Jar'

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Add Javadoc

yadobler committed:
Fix unused argument in method signature

yadobler committed:
Fix unused argument in method signature

yadobler committed:
Add google java checkstyle config to gradle

yadobler committed:
Fix gradlew pointing to wrong wrapper

yadobler committed:
Change checkstyle indent to 4

yadobler committed:
Export .editorconfig with style settings

yadobler committed:
Fix checkstyle violations

yadobler committed:
Fix checkstyle import order checks

yadobler committed:
Fix checkstyle import order check for SAME_PACKAGE

yadobler committed:
Fix checkstyle violations

yadobler committed:
Move TaskList related classes into tasks.tasklist package

yadobler committed:
Add isStringFoundInTask function for Task objects

yadobler committed:
Add new replies for Find function

yadobler committed:
Add 'find' command in help text

yadobler committed:
Create new TaskListFilterView extending TaskList

yadobler committed:
Update field members to protected

yadobler committed:
Implement method overrides for ad and delete

yadobler committed:
Add find function

yadobler committed:
Add 'find' and 'reset' command to yappingbot

yadobler committed:
Move adding task into tasklist from main loop to method in command class

yadobler committed:
Fix ditto commit intentions

yadobler committed:
Fix 'find' method substring including initial space

yadobler committed:
Fix bug - task array incorrectly declared as static

yadobler committed:
Fix find command and TaskListFilterView factory method issues

yadobler committed:
Add trimming to userinput string

yadobler committed:
Fix bug and mental health

yadobler committed:
Merge branch 'branch-A-JavaDoc'

yadobler committed:
Resolve merge conflict

yadobler committed:
Fix bugs

yadobler committed:
Resolve merge conflict

yadobler committed:
Fix checkstyle violations

yadobler committed:
Fix checkstyle import grouping

nigellqx commented on src/main/java/yappingbot/ui/Ui.java:
```diff
@@ -0,0 +1,106 @@
+
+    /**
+     * Decorates the given string, to denote that it is the bot's output.
+     * This accepts multiple lines
```
Minor nitpick - I think you miss a period here.

yadobler replied:
Ah how did checkstyle miss this...
Good catch!

nigellqx commented on src/main/java/yappingbot/tasks/tasklist/TaskListFilterView.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public void add(Task task) {
```
Perhaps a name like addTask for the method might be clearer?

nigellqx commented on src/main/java/yappingbot/tasks/tasklist/TaskListFilterView.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public Task delete(int index) throws YappingBotOobException {
```
Perhaps a name like deleteTask for the method might be clearer?

nigellqx commented on src/main/java/yappingbot/stringconstants/ReplyTextMessages.java:
```diff
@@ -0,0 +1,98 @@
+/**
+ * String constants of messages that the bot can use to reply.
+ */
+public final class ReplyTextMessages {
```
I like the thought of putting all the string constants that the bot can use to reply in a class 👍

nigellqx commented on src/main/java/yappingbot/tasks/Task.java:
```diff
@@ -0,0 +1,150 @@
+ */
+public abstract class Task {
+    private String taskName;
+    private boolean taskDone;
```
Perhaps isTaskDone might be a clearer name for a boolean value?

nigellqx reviewed (commented):
LGTM - just a few minor potential points that you might want to consider for the coding quality

yadobler commented:
@nigellqx I added PR Description, if you wanna review. Otherwise, thanks, I have implemented your suggestions and will push soon.

yadobler committed:
Resolve PR review suggestions by @nigellqx

Rachael-Chan reviewed (commented):
LGTM - Very clear naming of methods, classes and variables.

yadobler commented:
Thank you @Rachael-Chan and good night!

yadobler committed:
Update README

yadobler committed:
Update gradle for javafx

yadobler committed:
Fix nix-os library paths for javafx

yadobler committed:
Convert nix shell helper script to hardlink

yadobler committed:
Add basic javafx test

yadobler committed:
Change main entry to Launcher, with choice of GUI or CLI

yadobler committed:
Create GUI with fxml

yadobler committed:
Delete to prepare refactoring cli Ui as object

yadobler committed:
Split UI into UiCli and UiGui

yadobler committed:
Add reset text message

yadobler committed:
Create exception for IO problems

yadobler committed:
Encapsulate CLI ui in UiCli

yadobler committed:
Convert Ui to interface

yadobler committed:
Convert Ui to interface

yadobler committed:
Create UI interface in main before massing to yappingbot

yadobler committed:
Update visibility of methods

yadobler committed:
Update visibility of methods

yadobler committed:
Do many stuff while sleepy at 4am

yadobler committed:
Refactor we do a little bit of hmhmhm

yadobler committed:
Move tasklist view reset method to command dispatcher

yadobler committed:
Move launching logic to launcher; rename Main to MainGuiApplication

yadobler committed:
Fix accidental discard of resetView method return value

yadobler committed:
Update runtest.sh to for regression testing

yadobler committed:
Fix many bugs

yadobler committed:
Fix UI related bugs

yadobler committed:
Finish CLI related bugs

yadobler committed:
Add GUI implementation without logic

yadobler committed:
Add chat dialog update, without logic

yadobler committed:
Made it work. wow

yadobler committed:
Complete GUI integration

yadobler committed:
Merge branch 'branch-Level-10'

yadobler committed:
Update README

yadobler committed:
Change push to add for methods adding into I/O buffer in UiGui

All statements that adds Strings into input and output buffer pushes
them in the wrong order.

When polled, the buffer returns the queued strings in the wrong order,
as seen with `find` command - the list printout comes before the
confirmation reply.

All instances of `xBuffer.push` to be replaced with `xBuffer.add`
(where x is `input` or `output`).

This ensures the Strings are properly enqueued into the buffer to be
polled in order.

Note: ensure push is paired with pop, and add is paired with poll for
future uses of ArrayDeque.

yadobler committed:
Create gradle.yml

yadobler committed:
Merge pull request #1 from yadobler/branch-A-CI

Create gradle.yml

yadobler committed:
Update gradle.yml

yadobler committed:
Update gradle.yml to only build on all platform

Rest build on `ubuntu-latest`

yadobler committed:
Update gradle.yml to remove duplicate checkstyle run

yadobler committed:
Fix checkstyle violations and import order rule discrepancy

Code has various checkstyle violations and the order import did not
tally with the requirements, particularly javafx files were not
counted as a same group with same package files.

This prevents CI from running since code violations are a failure in
the gradlew build and shadowJar process.

Modified checkstyle.xml to follow AddressBook3 example, and fixed
violations reported by checkstyle.

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Update jUnit test files

jUnit test files are outdated after refactoring I/O out of bot and
implementing it as seperate interface.

This must be updated to ensure the regresion testing of the CLI
component of this bot.

Fixed the expected messages and converted static to object reference
for the UiCli tests.

yadobler committed:
Update gradle.yml to upload artifact generated by shadowJar

yadobler committed:
Update gradle.yml to build without clean and test

yadobler committed:
Update gradle.yml to clean and build but not test (done in separate job)

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Update gradle.yml to omit dependency graph updating, and combine jar release with testing and remove redundency

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Fix test strings to use manual string concat over triple-quotes

yadobler committed:
Update gradle.yml to generate changes between tags

yadobler committed:
Update gradle.yml fixing changelog step

yadobler committed:
Update gradle.yml replace unescaped colon with equals sign

yadobler committed:
Update gradle.yml

yadobler committed:
Update gradle.yml to just use tag name for release file

Gosh I'm in the toilet, this better work.

yadobler committed:
Move checkstyle testing to testAndGenerateReleaseFile

yadobler committed:
Split gradle.yml to 2 workflows, one for push to master, other for push tags to master

yadobler committed:
Add token for pushing release

yadobler committed:
Fix token for CI releasing

yadobler committed:
Add assertions to regions that assumes non-OOB or assumes non-Null

yadobler committed:
Extract collection of multiple exceptions into its own class

yadobler committed:
Standardize spelling of userInputSlices

yadobler committed:
Fix code standards

yadobler committed:
Add extra assertions

yadobler committed:
Fix code standards

yadobler committed:
Optimize inputs

yadobler committed:
Cleanup code

yadobler committed:
Merge pull request #2 from yadobler/branch-A-Assertions

Merge branch-A-Assertions into master

yadobler committed:
Merge branch 'master' into branch-A-CodeQuality

yadobler committed:
Fix typo from merge conflict resolution

yadobler committed:
Merge pull request #3 from yadobler/branch-A-CodeQuality

Merge branch-A-CodeQuality to master

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Create CommandBase to prepare refactoring commands out of CommandDispatch

yadobler committed:
Enable assertions in gradle

yadobler committed:
Commit before switching branch temporarily

yadobler committed:
Create CommandBase for commands to follow

yadobler committed:
Commit temporarily for swtich

yadobler committed:
Refine CommandBase constructor

yadobler committed:
Add Pair class for convenience use

yadobler committed:
Replace Pair import from javafx for new Pair class in package

yadobler committed:
Create ArgEnums interface to restrict CommandBase

yadobler committed:
Migrate Launcher logic to LauncherCommand class

yadobler committed:
Refactor Commands to not require pair but directly parse

yadobler committed:
Complete migrating Launcher logic to LauncherClass

yadobler committed:
Migrate resetView method into new class

yadobler committed:
Migrate findStringInTasks method into new class

yadobler committed:
Migrate printUserList method into new class

yadobler committed:
Add check for fulfilling number of arguments

yadobler committed:
Migrate deleteTask method into new class

yadobler committed:
Migrate createTask method into new class

yadobler committed:
Migrate createTask Todo, Event, Deadline methods into new class

yadobler committed:
Move CommandTypes enum into Parser

yadobler committed:
Delete redundant CommandDispatcher

yadobler committed:
Finalize command abstraction

yadobler committed:
Add easier alises for commands

yadobler committed:
Add alias command

yadobler committed:
Merge branch 'branch-C-FriendlierSyntax'

yadobler committed:
Fix LauncherCommand bug

yadobler committed:
Update README.md

yadobler committed:
Add css styling

yadobler committed:
Embody I-am-a-designer spirit

yadobler committed:
Merge branch 'branch-A-BetterGui'

GUI is ugly.

Add CSS stylesheet to beautify the GUI.

Gui is now pretty.

Colors are horrible use of TokyoNight dark high-saturation.

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Add user guide

yadobler committed:
Fix css bug

yadobler committed:
Add ui image

yadobler committed:
Add image to readme

yadobler committed:
Merge branch 'branch-A-UserGuide'

yadobler committed:
Update README.md formatting issue

yadobler committed:
Switch to docs/Ui.png for image

yadobler committed:
Delete Ui.png duplicate in root

yadobler committed:
Update gradle.yml to do smoke testing

yadobler committed:
Update gradle.yml to do smoke test for all platforms

yadobler committed:
Revert gradle.yml smoke test

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Add _config.yaml

yadobler committed:
Merge pull request #5 from yadobler/branch-A-UserGuide

Add _config.yaml

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Add custom html

yadobler committed:
Fix release url

yadobler committed:
Fix release url 2

yadobler committed:
Fix format typo

yadobler committed:
Fix hline issues in README.md

yadobler committed:
Add myself to Contributors

yadobler committed:
Add unit test for GUI ui, both sync and async

yadobler committed:
Add proper newline wrapping

yadobler committed:
Add jUnit test for Todo task

yadobler committed:
Add jUnit test for Deadline task

yadobler committed:
Add jUnit test for Event task

yadobler committed:
Fix jUnit test for Deadline task deserialisation

yadobler committed:
Add jUnit test for TaskList

yadobler committed:
Add jUnit test for TaskListFilterView

yadobler committed:
Merge branch 'branch-A-MoreTesting'

yadobler committed:
Merge branch 'master' of https://github.com/yadobler/ip

yadobler committed:
Fix wrap text issue

yadobler committed:
Fix alignment issue in dialog box

yadobler committed:
Fix UiCli wrapping issue

yadobler committed:
Add helper method to concat argument value array into String

yadobler committed:
Fix incorrect argument value checking

yadobler committed:
Fix uncatched exception thrown

yadobler committed:
Add CLI smoketest in CI

yadobler committed:
Add runtest.sh to CI for testing expected input and output

yadobler committed:
Sync IntelliJ run configs

yadobler committed:
Fix checkstyle violations

yadobler committed:
Fix code violations

yadobler committed:
Add code violation checker script

yadobler committed:
Merge branch 'branch-bugFixing'

yadobler committed:
Add -p switch to output offending lines

yadobler committed:
Comply with issue #2

yadobler committed:
Add stuff and things

yadobler committed:
Fix regex checks for violation

yadobler committed:
Add new UI images
